# 課題2 データの集計

## Q1
中央値はデータの中心傾向を観察するための記述統計の指標です。同じくデータの中心傾向の指標である平均値と比べると、データの外れ値の影響を受けにくという特徴があります。1変数の$n$個のデータを考えた時、中央値を求めるにはまず、データの値の大きさの順に並び替えます。次に、データの数が奇数個の場合は、大きさの順に並び替えた$(n+1)/2$番目の値、データの数が偶数個の場合は、大きさの順に並び替えた$n/2$番目と$n/2+1$の値の平均の値、をそれぞれデータの中央値とします。

分散はデータのばらつきを観察するための記述統計の指標です。個々のデータがデータ全体の平均からどらぐらいばらついているかを表します。以下の通り、個々のデータの値$x_{i}$とデータ全体の平均の値$\overline{x}$の差の2乗和をデータ数$n$で割ったものが分散になります。

$\frac{1}{n}\Sigma^{n}_{i=1}(x_{i}-\overline{x})^{2}$

0から100までの整数を要素とする任意の長さのリストを引数で入力として受け取り、以下を要素とするリストを返す`avg_med_var()`関数を完成させてください。
- 入力リストの全ての要素の平均値
- 入力リストの全ての要素の中央値
- 入力リストの全ての要素の分散

In [17]:
def avg_med_var(input_list):
    mean_value=0 #平均値を初期化
    med_value=0 # 中央値を初期化
    var_value=0 # 分散を初期化
    results=[] # 結果のリスト
        
    input_list.sort() # リストの要素を昇順に並び替え
    mid_idx = len(input_list)//2 # リストの中央（長さの半分）のインデックス（必要であれば利用する）
    if len(input_list) % 2 == 0:
        med_value= (input_list[mid_idx] + input_list[mid_idx-1])/2
    else:
        med_value = input_list[mid_idx]
    
    mean_value = sum(input_list)/len(input_list) #平均値
    mse_value = 0 # データの値と平均値の差の2乗和を初期化
    for i in range(len(input_list)):
        mse_value += (mean_value-input_list[i])**2 
    var_value = mse_value/len(input_list)
    
    results.append(mean_value) 
    results.append(med_value) 
    results.append(var_value)
    return results   

`avg_med_var()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [18]:
print(avg_med_var([1,4,3,2,5,6,8,7,9,10]))
print(avg_med_var([1,4,3,2,5,6,8,7,9,10,11]))

[5.5, 5.5, 8.25]
[6.0, 6, 10.0]


以下のセルを実行して、実行結果が上記と同じであること確認してください。結果が異なる場合は、作成した`avg_mean_var`関数に誤りがないか確認してください。

In [19]:
# avg_med_var関数の結果を検証するための関数
import numpy as np
def avg_med_var_test(input_list):
    return [np.mean(input_list), np.median(input_list), np.var(input_list)]

In [20]:
print(avg_med_var_test([1,4,3,2,5,6,8,7,9,10]))
print(avg_med_var_test([1,4,3,2,5,6,8,7,9,10,11]))

[5.5, 5.5, 8.25]
[6.0, 6.0, 10.0]


In [21]:
# !!!このセルは変更しないでください!!! 
assert avg_med_var([1,4,3,2,5,6,8,7,9,10]) == avg_med_var_test([1,4,3,2,5,6,8,7,9,10])

In [22]:
# !!!このセルは変更しないでください!!! 
assert avg_med_var([1,4,3,2,5,6,8,7,9,10,11]) == avg_med_var_test([1,4,3,2,5,6,8,7,9,10,11])

# Q2
以下のような形式の"user_score.csv"ファイルを読み込み、データフレーム`score`を作成します。
```Python
## user_score.csvファイル
user, kokugo, shakai, sugaku, rika
1, 30, 43, 51, 63
2, 39, 21, 49, 56
...
```

In [23]:
import pandas as pd
score = pd.read_csv('user_score.csv')
score.head(5)

user  kokugo  shakai  sugaku  rika
0     1      30      43      51    63
1     2      39      21      49    56
2     3      29      30      23    57
3     4      95      87      77   100
4     5      70      71      78    67

データフレームの`desribe()`メソッドを使って、各教科ごとの記述統計を確認してください

In [24]:
score.describe()

user      kokugo      shakai      sugaku        rika
count  166.000000  166.000000  166.000000  166.000000  166.000000
mean    83.500000   52.331325   39.602410   45.614458   49.885542
std     48.064193   21.835858   21.456335   24.236127   21.408130
min      1.000000    0.000000    4.000000    0.000000    9.000000
25%     42.250000   35.000000   23.000000   28.000000   31.000000
50%     83.500000   53.000000   35.000000   43.000000   49.000000
75%    124.750000   70.000000   53.750000   63.750000   65.750000
max    166.000000   96.000000   90.000000  100.000000  100.000000

### Q2-1
データフレーム`score`から各教科の得点がすべて平均点以上であるようなユーザの数を返す`above_mean()`関数を作成してください

In [27]:
def above_mean(df):
    return len(df[(df['kokugo'] > df['kokugo'].mean()) & (df['shakai'] > df['shakai'].mean()) & (df['sugaku'] > df['sugaku'].mean()) & (df['rika'] > df['rika'].mean())])

`above_mean()`関数が完成したら、以下のセルを実行して動作を確認してください。条件満たすユーザは46人となります。

In [28]:
above_mean(score)

46

In [29]:
### このセルは変更しないでください ###
assert 46 == above_mean(score)

### Q2-2
ユーザごとの全教科の得点の総和の値を表す'sum'列をデータフレーム`score`に新たに追加したデータフレームを返す`score_sum()`関数を作成してください。以下のようにして、データフレームの指定した列の行方向の値の和を持つシリーズオブジェクト作ることができます。
```Python
データフレーム[[列1,列2,...,]].sum(axis=1)
```

In [40]:
def score_sum(df):
    sum_list = []
    for i, j in df.iterrows():
        sum_list.append(j['kokugo'] + j['sugaku'] + j['rika'] + j['shakai'])
    df['sum'] = sum_list
    return df

In [41]:
score_sum(score)

user  kokugo  shakai  sugaku  rika  sum
0       1      30      43      51    63  187
1       2      39      21      49    56  165
2       3      29      30      23    57  139
3       4      95      87      77   100  359
4       5      70      71      78    67  286
5       6      67      53      56    61  237
6       7      29      26      44    52  151
7       8      56      54      37    59  206
8       9      45      21       7    44  117
9      10      68      41      29    81  219
10     11      50      43      80    73  246
11     12      70      61      61    71  263
12     13      46      37      36    56  175
13     14      23      24      20    36  103
14     15      77      54      58    84  273
15     16      15       8      22    29   74
16     17      37      23      23    32  115
17     18      55      42      85    86  268
18     19       0       5       0    24   29
19     20      35      48      24    42  149
20     21      60      80      64    72  276
21     22      70      52      75    60  257
22     23      82      84      86    88  340
23     24      44      22      18    13   97
24     25      54      14      33    43  144
25     26      80      63      59    73  275
26     27      46      47      72    56  221
27     28      48      63      72    83  266
28     29      51      41      63    60  215
29     30      37      18      38    51  144
..    ...     ...     ...     ...   ...  ...
136   137      53      23      28    30  134
137   138      70      41      44    34  189
138   139      78      53      45    54  230
139   140      39      19       5    32   95
140   141      74      62      44    58  238
141   142      33      33      32    42  140
142   143      43      24      29    47  143
143   144      54      21      43    28  146
144   145      26      15      13    19   73
145   146      58      17      34    19  128
146   147      29      15      37    19  100
147   148      73      60      59    54  246
148   149      59      25      45    48  177
149   150      58      17      31    23  129
150   151      70      56      76    62  264
151   152      19      12       9    28   68
152   153      29      21      26    30  106
153   154      32       7      14    18   71
154   155      77      64      87    71  299
155   156      91      69     100    83  343
156   157      50      31      53    44  178
157   158      63      32      43    52  190
158   159      42      16      27    22  107
159   160      68      28      60    69  225
160   161      70      23      43    50  186
161   162      82      78      80    88  328
162   163       0       8       2     9   19
163   164      45      26      29    24  124
164   165      73      31      43    32  179
165   166      60      85      89    80  314

[166 rows x 6 columns]

`score_sum()`関数が完成したら、以下のセルを実行して動作を確認してください。'sum'列が追加されていることを確認してください。

In [42]:
score = score_sum(score)
score.head(5)

user  kokugo  shakai  sugaku  rika  sum
0     1      30      43      51    63  187
1     2      39      21      49    56  165
2     3      29      30      23    57  139
3     4      95      87      77   100  359
4     5      70      71      78    67  286

全教科の合計得点の上位3人の'user'を1位、2位、3位の順に要素として持つリストを返す`score_top3()`関数を作成させてください。

In [45]:
def score_top3(df):
    sorted_df = df.sort_values(['sum'], ascending=False)
    results=[]
    results.append(sorted_df.iloc[0]['user'])
    results.append(sorted_df.iloc[1]['user'])
    results.append(sorted_df.iloc[2]['user'])
    return results

`score_top3()`関数が完成したら、以下のセルを実行して動作を確認してください。`[39, 4, 50]`の'user'が合計得点の上位3人です。

In [46]:
score_top3(score)

[39, 4, 50]

In [47]:
### このセルは変更しないでください ###
assert [39, 4, 50] == score_top3(score)

### Q2-3
データフレーム`score`の'sum'列を対象に、以下を階級幅として各階級の度数を要素として持つリストを返すscore_hist()関数を完成させてください。
- 0以上、50以下
- 51以上、100以下
- 101以上、150以下
- 151以上、200以下
- 201以上、250以下
- 251以上、300以下
- 301以上、350以下
- 351以上、400以下

In [61]:
def score_hist(df):
    counter0 = 0
    counter1 = 0
    counter2 = 0
    counter3 = 0
    counter4 = 0
    counter5 = 0
    counter6 = 0
    counter7 = 0
    for i, j in df.iterrows():
        if 0 <= j['sum'] <51:
            counter0 += 1
        if 51 <= j['sum'] <101:
            counter1 += 1
        if 101 <= j['sum'] <151:
            counter2 += 1
        if 151 <= j['sum'] <201:
            counter3 += 1
        if 201 <= j['sum'] <251:
            counter4 += 1
        if 251 <= j['sum'] <301:
            counter5 += 1
        if 301 <= j['sum'] <351:
            counter6 += 1
        if 351 <= j['sum'] <401:
            counter7 += 1
    results = [counter0, counter1, counter2, counter3, counter4, counter5, counter6, counter7]
    return results

`score_hist()`関数が完成したら、以下のセルを実行して動作を確認してください。`[4, 20, 39, 35, 29, 24, 13, 2]`が'sum'の度数分布です。

In [62]:
score_hist(score)

[4, 20, 39, 35, 29, 24, 13, 2]

In [63]:
### このセルは変更しないでください ###
assert [4, 20, 39, 35, 29, 24, 13, 2] == score_hist(score)

### Q2-4
以下のような形式の"user_class.csv"ファイルを読み込み、データフレーム`user_class`を作成します。
```Python
## user_class.csvファイル
'user', 'class'
1, 'C'
2, 'C'
3, 'B'
...
```

In [64]:
user_class = pd.read_csv('user_class.csv')
user_class.head(5)

user class
0     1     C
1     2     C
2     3     B
3     4     C
4     5     A

'user'の列の値をキーに、2つのデータフレーム、user_classとscore、を結合して新しいデータフレームを作成した上で、データフレームの`groupby`メソッドを使って、'class'をインデックス、'class'ごとの'sum'の最高値を列として持つシリーズオブジェクトを返す`score_by_class()`関数を完成させてください。

In [82]:
def score_by_class(df1, df2):
    ### user_classとscoreを結合して新しいデータフレームを作成するコード ###
    df_merged = pd.merge(df1, df2, on="user").sort_values(['sum'], ascending=False)
    class_top = df_merged.groupby('class')["class", "sum"].head(1)
    sum_list = []
    index = []
    for i, j in class_top.iterrows():
        sum_list.append(j['sum'])
        index.append(j['class'])
    series = pd.Series(sum_list, index=index)
    return series

In [83]:
print(score_by_class(user_class, score))

A    361
C    359
B    349
dtype: int64


`score_by_class()`関数が完成したら、以下のセルを実行して動作を確認してください。`{'A': 361, 'B': 349, 'C': 359}`が'class'ごとの'sum'の最高値です。

In [84]:
score_by_class(score, user_class).to_dict()    

{'A': 361, 'B': 349, 'C': 359}

In [85]:
assert {'A': 361, 'B': 349, 'C': 359} == score_by_class(score, user_class).to_dict()